conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/bitcoin/BitcoinOperations.scala

```diff
@@ -0,0 +1,100 @@
+import slickeffect.Transactor
+
+import tech.cryptonomic.conseil.common.config.Platforms.BitcoinBatchFetchConfiguration
+import tech.cryptonomic.conseil.common.rpc.RpcClient
+import tech.cryptonomic.conseil.common.bitcoin.{BitcoinPersistence, Tables}
+import tech.cryptonomic.conseil.common.bitcoin.rpc.BitcoinClient
+import tech.cryptonomic.conseil.indexer.config.{Custom, Depth, Everything, Newest}
+
+/**
+  * Bitcoin operations for Lorre.
+  *
+  * @param bitcoinClient Bitcoin client instance
+  * @param persistence Bitcoin persistence instance
+  * @param tx [[slickeffect.Transactor]] to perform a Slick operations on the database
+  */
+class BitcoinOperations[F[_]: Concurrent](
+    bitcoinClient: BitcoinClient[F],
+    persistence: BitcoinPersistence[F],
+    tx: Transactor[F],
+    batchConf: BitcoinBatchFetchConfiguration
+) extends LazyLogging {
+
+  /**
+    * Start Lorre with mode defined with [[Depth]].
+    *
+    * @param depth Can be: Newest, Everything or Custom
+    */
+  def loadBlocks(depth: Depth): Stream[F, Unit] =
+    Stream
+      .eval(tx.transact(Tables.Blocks.sortBy(_.height).take(1).result)) // get last block height
+      .zip(bitcoinClient.getBlockChainInfo)
+      .evalTap(_ => Concurrent[F].delay(logger.info(s"Start Lorre for Bitcoin")))
+      .flatMap {
+        case (block, info) =>
+          depth match {
+            case Newest if block.size > 0 => loadBlocksWithTransactions(block.head.height to info.blocks)
+            case Newest | Everything => loadBlocksWithTransactions(1 to info.blocks)
+            case Custom(depth) => loadBlocksWithTransactions(depth to info.blocks)
+          }
+      }
+
+  /**
+    * Get Blocks from Bitcoin node through Bitcoin client and save them into the database using Slick.
+    * In the beginning, the current list of blocks is obtained from the database and removed from the computation.
+    *
+    * @param range Inclusive range of the block's height
+    */
+  def loadBlocksWithTransactions(range: Range.Inclusive): Stream[F, Unit] =
+    Stream
+      .eval(getExistingBlocks(range))
+      .flatMap(
+        existingBlocks =>
+          Stream
+            .range(range.start, range.end)
+            .filter(height => !existingBlocks.contains(height))
+            .through(bitcoinClient.getBlockHash(batchConf.hashBatchSize))
+            .through(bitcoinClient.getBlockByHash(batchConf.blocksBatchSize))
+            .through(bitcoinClient.getBlockWithTransactions(batchConf.transactionsBatchSize))
+            .evalTap { case (block, _) => Concurrent[F].delay(logger.info(s"Save block with height: ${block.height}")) }
+            .map((persistence.createBlock _).tupled)
+            .evalMap(tx.transact)
+            .drain
+      )
+
+  /**
+    * Get sequence of existing blocks from the database.
+    *
+    * @param range Inclusive range of the block's height
+    * @return
+    */
+  def getExistingBlocks(range: Range.Inclusive): F[Seq[Int]] =
```

Comment:
I believe we need this method (instead of fetching the latest) due to concurrent nature of RPC Client? (Just to confirm). Asking, cause this list might be pretty long, one we will have million of blocks, right? Which specific implementation is hidden under `Seq` here - it will make a difference in long run I guess.

Reply:
Can it be renamed to `getIndexedBlockHeights`

To differentiate from "remote-fetching" and to make clear what is the return type?